File: C#/PBTKata/Rentals/RentalCalculator.cs
using System;
using System.Text;
using LanguageExt;
using static LanguageExt.Prelude;

namespace PBTKata.Rentals
{
    public class RentalCalculator
    {
        public List<Rental> Rentals { get; init; }
        public bool Calculated { get { return calculated; } }
        public double Amount { get { return amount; } }

        private double amount = 0;
        private bool calculated = false;

        public RentalCalculator(List<Rental> rentals)
        {
            Rentals = rentals;
        }

        public Either<String, String> CalculateRental()
        {
            if(Rentals.IsNull() || Rentals.Count == 0)
            {
                return Left("No rentals !!!");
            }
            else
            {
                var result = new StringBuilder();

                foreach (var rental in Rentals)
                {
                    if(!calculated)
                    {
                        this.amount += rental.Amount;
                        result.AppendLine(FormatLine(rental, amount));
                    }
                }
                result.AppendLine($"Total amount | {this.amount:N2}");
                calculated = true;

                return Right(result.ToString());
            }
        }

        private static String FormatLine(Rental rental, Double amount) =>
            $"{rental.Date} : {rental.Label} | {rental.Amount:N2}";
    }
}

File: java/src/main/java/rentals/RentalCalculator.java
package rentals;

import lombok.Getter;

import java.util.List;

@Getter
public class RentalCalculator {
    private final List<Rental> rentals;
    private double amount;
    private boolean calculated;

    public RentalCalculator(List<Rental> rentals) {
        this.rentals = rentals;
    }

    public String calculateRental() {
        if (rentals.isEmpty()) {
            throw new IllegalStateException("No rentals !!!");
        }

        var result = new StringBuilder();

        for (var rental : rentals) {
            if (!calculated)
                this.amount += rental.amount();

            result.append(formatLine(rental, amount));
        }
        result.append(String.format("Total amount | %f", this.amount));
        calculated = true;

        return result.toString();
    }

    private String formatLine(Rental rental, double amount) {
        return String.format("%tF : %s | %f \n",
                rental.date(),
                rental.label(),
                rental.amount());
    }
}